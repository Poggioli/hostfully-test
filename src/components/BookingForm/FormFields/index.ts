export * from './FormFieldDateRange'
export * from './FormFieldQuantityGuests'
