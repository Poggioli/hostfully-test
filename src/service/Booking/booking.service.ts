/* eslint-disable @typescript-eslint/no-explicit-any */
import { useStore } from "@/lib/store";
import axiosClient from "@/service/api";
import { useMutation, useQuery, useQueryClient } from "react-query";
import { Booking } from "./booking.types";

const keys = {
  getBookings: ["bookings"],
  deleteBookings: ["deleteBookings"],
  updateBookings: ["updateBookings"],
  postBookings: ["postBookings"],
};

export function useGetBooking(options?: any) {
  const { setBookings } = useStore();

  return useQuery<Booking[]>(keys.getBookings, {
    queryFn: async () => {
      const res = await axiosClient.get<Booking[]>("/bookings");
      return res.data;
    },
    onSuccess(data) {
      setBookings(data);
    },
    ...options,
  });
}

export function useDeleteBooking(options?: any) {
  const queryClient = useQueryClient();
  const { deleteBooking, addBooking } = useStore();

  return useMutation(
    keys.deleteBookings,
    async (id: string, booking?: Booking) => {
      await axiosClient.delete(`/bookings/${id}`);
      return booking;
    },
    {
      onMutate: async (id: string) => {
        await queryClient.cancelQueries({ queryKey: keys.deleteBookings });
        const previousBookings = queryClient.getQueryData(keys.getBookings);
        deleteBooking(id);
        return { previousBookings };
      },
      onError: (err, variables, context: any) => {
        const booking: Booking = context.previousBookings as Booking;
        addBooking(booking);
      },
      onSettled: () => {
        queryClient.invalidateQueries({ queryKey: keys.getBookings });
      },
      ...options,
    }
  );
}

export function useEditBooking(options?: any) {
  const queryClient = useQueryClient();
  const { editBooking } = useStore();

  return useMutation(
    keys.updateBookings,
    async (booking: Booking) => {
      await axiosClient.put(`/bookings/${booking.id}`, booking);
      return booking;
    },
    {
      onMutate: async (booking: Booking) => {
        await queryClient.cancelQueries({ queryKey: keys.updateBookings });
        const previousBookings = queryClient.getQueryData<Booking[]>(
          keys.getBookings
        ) as Booking[];
        editBooking(booking);
        return {
          previousBooking: previousBookings.find((b) => b.id === booking.id),
        };
      },
      onError: (err, variables, context: any) => {
        const booking: Booking = context.previousBooking as Booking;
        editBooking(booking);
      },
      onSettled: () => {
        queryClient.invalidateQueries({ queryKey: keys.getBookings });
      },
      ...options,
    }
  );
}

export function usePostBooking(options?: any) {
  const queryClient = useQueryClient();
  const { deleteBooking, addBooking } = useStore();

  return useMutation(
    keys.postBookings,
    async (booking: Booking) => {
      const res = await axiosClient.post(`/bookings`, booking);
      return res.data;
    },
    {
      onMutate: async (booking: Booking) => {
        await queryClient.cancelQueries({ queryKey: keys.postBookings });
        const previousBookings = queryClient.getQueryData<Booking[]>(
          keys.getBookings
        ) as Booking[];
        addBooking(booking);
        return { previousBookings };
      },
      onError: (err, variables, context: any) => {
        const booking: Booking = context.previousBooking as Booking;
        deleteBooking(booking.id);
      },
      onSettled: () => {
        queryClient.invalidateQueries({ queryKey: keys.getBookings });
      },
      ...options,
    }
  );
}
