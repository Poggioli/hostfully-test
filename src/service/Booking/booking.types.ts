export type Booking = {
  id: string;
  roomId: string;
  totalPrice: number;
  checkIn: Date;
  checkOut: Date;
  quantityGuests: number;
}