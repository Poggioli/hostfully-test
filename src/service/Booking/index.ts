export * from './booking.service';
export * from './booking.types';
