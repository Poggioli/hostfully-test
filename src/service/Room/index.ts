export * from './room.service';
export * from './room.types';

