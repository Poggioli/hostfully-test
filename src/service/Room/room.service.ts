/* eslint-disable @typescript-eslint/no-explicit-any */
import { useStore } from "@/lib/store";
import { useQuery } from "react-query";
import { Room } from ".";
import axiosClient from "../api";

const keys = {
  getRooms: ["rooms"],
};

export function useGetRooms(options?: any) {
  const { setRooms } = useStore();

  return useQuery<Room[]>(keys.getRooms, {
    queryFn: async () => {
      const res = await axiosClient.get<Room[]>("/rooms");
      return res.data;
    },
    onSuccess(data: any) {
      setRooms(data);
    },
    ...options,
  });
}
