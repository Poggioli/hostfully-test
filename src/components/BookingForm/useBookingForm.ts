import { useBookingDrawer } from "@/components/BookingDrawer";
import { useStore } from "@/lib/store";
import { Booking, useEditBooking, usePostBooking } from "@/service/Booking";
import { Room } from "@/service/Room";
import { zodResolver } from "@hookform/resolvers/zod";
import {
  addDays,
  areIntervalsOverlapping,
  differenceInCalendarDays,
  differenceInDays,
  eachDayOfInterval,
  isBefore,
} from "date-fns";
import { useMemo } from "react";
import { useForm } from "react-hook-form";
import { v4 as uuidv4 } from "uuid";
import * as z from "zod";

const TODAY = new Date();
TODAY.setHours(0, 0, 0, 0);
const MIN_CHECKOUT = addDays(new Date(), 1);

const bookingSchema = z
  .object({
    checkIn: z
      .date({
        required_error: "This field is required",
      })
      .min(TODAY, {
        message: "Select only days from today.",
      }),
    checkOut: z
      .date({
        required_error: "This field is required",
      })
      .min(MIN_CHECKOUT),
    quantityGuests: z
      .number({
        required_error: "This field is required",
        coerce: true,
        invalid_type_error: "This field is required",
      })
      .min(1),
  })
  .superRefine((val, ctx) => {
    const { checkIn, checkOut } = val;
    if (isBefore(checkOut, checkIn)) {
      ctx.addIssue({
        path: ["checkOut"],
        code: z.ZodIssueCode.invalid_date,
        message: "Checkout must be after check in",
      });
      return;
    }

    if (differenceInDays(checkIn, checkOut) === 0) {
      ctx.addIssue({
        path: ["checkOut"],
        code: z.ZodIssueCode.invalid_date,
        message: "The booking must be for at least 1 day",
      });
      return;
    }
  });

export type BookingSchema = z.infer<typeof bookingSchema>;

type UseBookingFormProps = {
  room: Room;
  booking?: BookingSchema & {
    id: string
  }
};

export const useBookingForm = ({ room, booking }: UseBookingFormProps) => {
  const { onClose } = useBookingDrawer();
  const { getBookingsByIdRoom } = useStore();
  const { mutate: createBooking } = usePostBooking();
  const { mutate: editBooking } = useEditBooking();
  const bookings = getBookingsByIdRoom(room.id);

  const form = useForm<BookingSchema>({
    resolver: zodResolver(bookingSchema),
    defaultValues: {
      ...(booking && ({
        ...booking,
        checkIn: new Date(booking.checkIn),
        checkOut: new Date(booking.checkOut)
      }))
    }
  });

  const unavailableDates: Date[] = bookings.reduce(
    (prev: Date[], curr: Booking) => {
      const datesInInterval = curr.id !== booking?.id ? eachDayOfInterval({
        start: curr.checkIn,
        end: curr.checkOut,
      }) : [];
      return [...prev, ...datesInInterval];
    },
    []
  );

  const bookingDays: number = useMemo(() => {
    return (
      Math.abs(
        differenceInCalendarDays(form.watch("checkIn"), form.watch("checkOut"))
      ) || 0
    );
  }, [form.watch("checkIn"), form.watch("checkOut")]);

  const totalPriceGuest: number = useMemo(() => {
    return (form.watch("quantityGuests") || 0) * room.pricePerGuest;
  }, [form.watch("quantityGuests"), room.pricePerGuest]);

  const totalPriceDay: number = useMemo(() => {
    return room.pricePerDay * bookingDays;
  }, [room.pricePerDay, bookingDays]);

  const totalPrice: number = useMemo(() => {
    return totalPriceDay + totalPriceGuest;
  }, [totalPriceDay, totalPriceGuest]);

  const onSubmit = (data: BookingSchema) => {
    const isOverlapping = bookings.some((b) =>
      b.id !== booking?.id ? areIntervalsOverlapping(
        { start: data.checkIn, end: data.checkOut },
        { start: b.checkIn, end: b.checkOut },
        { inclusive: true }
      ) : false
    );

    if (isOverlapping) {
      form.setError("root", {
        type: "validate",
        message: "There are days that are already busy in your reservation.",
      });
      return;
    }

    if (booking) {
      editBooking({
        id: booking.id,
        roomId: room.id,
        totalPrice: totalPrice,
        ...data,
      });
      onClose();
      return;
    }

    createBooking({
      id: uuidv4(),
      roomId: room.id,
      totalPrice: totalPrice,
      ...data,
    });

    onClose();
  };

  return {
    form,
    onSubmit,
    totalPriceGuest,
    totalPriceDay,
    bookingDays,
    totalPrice,
    unavailableDates,
  };
};
