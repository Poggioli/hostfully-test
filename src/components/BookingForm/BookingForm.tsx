import { Button } from "@/components/ui/button";
import { Form, FormMessage } from "@/components/ui/form";
import { Room } from "@/service/Room";
import { FC } from "react";
import { useBookingDrawer } from "@/components/BookingDrawer";
import { CheckoutInfo } from "./CheckoutInfo";
import { FormFieldCheckin, FormFieldCheckout, FormFieldQuantityGuests } from "./FormFields";
import { useBookingForm } from "./useBookingForm";
import { Booking } from "@/service/Booking";

type BookingFormProps = {
    room: Room;
    booking?: Booking
}

const BookingForm: FC<BookingFormProps> = ({ room, booking }) => {
    const {
        form,
        onSubmit,
        totalPriceGuest,
        totalPriceDay,
        bookingDays,
        totalPrice,
        unavailableDates
    } = useBookingForm({ room, booking });
    const { onClose } = useBookingDrawer();

    return (
        <>
            <Form {...form}>
                <form onSubmit={form.handleSubmit(onSubmit)} className="flex flex-col gap-4">
                    <FormFieldCheckin unavailableDates={unavailableDates} />
                    <FormFieldCheckout unavailableDates={unavailableDates} />
                    <FormFieldQuantityGuests />
                    {form.formState.errors.root ? (
                        <FormMessage>{form.formState.errors.root.message}</FormMessage>
                    ) : null}
                    <CheckoutInfo
                        pricePerDay={room.pricePerDay}
                        pricePerGuest={room.pricePerGuest}
                        bookingDays={bookingDays}
                        quantityGuests={form.getValues('quantityGuests') || 0}
                        totalPrice={totalPrice}
                        totalPriceDay={totalPriceDay}
                        totalPriceGuest={totalPriceGuest}
                    />
                    <Button type="submit">{booking ? 'Save' : 'Book'}</Button>
                </form>
            </Form>
            <Button className="mt-4 w-full" onClick={onClose} variant="outline">Cancel</Button>
        </>
    )
};

export { BookingForm };
