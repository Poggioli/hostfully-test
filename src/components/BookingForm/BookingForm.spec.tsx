import { Booking } from "@/service/Booking";
import { Room } from "@/service/Room";
import { act, fireEvent, render, waitFor } from "@testing-library/react";
import { BookingForm } from "./BookingForm";

const mockOnClose = jest.fn();
jest.mock("@/components/BookingDrawer", () => ({
  useBookingDrawer: () => ({
    onClose: mockOnClose,
  }),
}));

const room: Room = {
  id: 'id-room-1',
  description: 'description room 1',
  name: 'name room 1',
  photos: [],
  pricePerDay: 400,
  pricePerGuest: 50
}

const booking: Booking = {
  id: "booking-1",
  roomId: room.id,
  totalPrice: 5000,
  checkIn: new Date(2025, 8, 20),
  checkOut: new Date(2025, 8, 25),
  quantityGuests: 3,
  photos: room.photos,
  roomName: room.name
}

const bookingToOverlap: Booking = { ...booking, id: "booking-2" }

const mockGetBookingsByIdRoom = jest.fn().mockReturnValue([booking]);

jest.mock("@/lib/store", () => {
  return {
    useStore: () => ({
      getBookingsByIdRoom: mockGetBookingsByIdRoom
    }),
  };
});

const mockPostMutate = jest.fn();
const mockPutMutate = jest.fn();
jest.mock("@/service/Booking", () => ({
  usePostBooking: () => ({ mutate: mockPostMutate }),
  useEditBooking: () => ({ mutate: mockPutMutate }),
}));

describe('BookingForm', () => {
  it(`GIVEN a open BookingForm
      WHEN user click in "Cancel"
      THEN should call onClose of useBookingDrawer`, () => {
    const { getByText } = render(<BookingForm room={room} booking={booking} />);
    const cancelButton = getByText('Cancel')

    act(() => fireEvent.click(cancelButton));
    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it(`GIVEN a BookingForm
        WHEN it's edit
        THEN should bring all fields filled`, () => {
    const { baseElement } = render(<BookingForm room={room} booking={booking} />);
    expect(baseElement).toMatchSnapshot();
  });

  it(`GIVEN a BookingForm
      WHEN it's create
      THEN should NOT bring fields filled`, () => {
    const { baseElement } = render(<BookingForm room={room} />);
    expect(baseElement).toMatchSnapshot();
  });

  describe('ERROR', () => {
    it(`GIVEN a form submitted
        WHEN its overlapping
        THEN should show overlap msg`, async () => {
      mockGetBookingsByIdRoom.mockReturnValue([booking, bookingToOverlap]);
      const { getByRole, getByText } = render(<BookingForm room={room} booking={bookingToOverlap} />);
      const submitButton = getByRole('button', {
        name: 'Save'
      });

      act(() => fireEvent.click(submitButton));

      await waitFor(() => {
        const errorRootMessage = getByText("There are days that are already busy in your reservation.");
        expect(errorRootMessage).toBeDefined()
      })

    });
  });
});