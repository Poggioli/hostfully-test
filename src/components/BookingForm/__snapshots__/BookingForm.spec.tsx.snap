// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingForm GIVEN a BookingForm
        WHEN it's edit
        THEN should bring all fields filled 1`] = `
<body>
  <div>
    <form
      class="flex flex-col gap-4"
    >
      <div
        class="space-y-2 flex flex-col"
      >
        <label
          class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 w-fit"
          for=":r5:-form-item"
        >
          Check-in
        </label>
        <button
          aria-controls="radix-:r6:"
          aria-describedby=":r5:-form-item-description"
          aria-expanded="false"
          aria-haspopup="dialog"
          aria-invalid="false"
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 w-[240px] pl-3 text-left font-normal"
          data-state="closed"
          id=":r5:-form-item"
          type="button"
        >
          September 20th, 2025
          <svg
            class="lucide lucide-calendar ml-auto h-4 w-4 opacity-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="18"
              rx="2"
              ry="2"
              width="18"
              x="3"
              y="4"
            />
            <line
              x1="16"
              x2="16"
              y1="2"
              y2="6"
            />
            <line
              x1="8"
              x2="8"
              y1="2"
              y2="6"
            />
            <line
              x1="3"
              x2="21"
              y1="10"
              y2="10"
            />
          </svg>
        </button>
      </div>
      <div
        class="space-y-2 flex flex-col"
      >
        <label
          class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 w-fit"
          for=":r7:-form-item"
        >
          Check-out
        </label>
        <button
          aria-controls="radix-:r8:"
          aria-describedby=":r7:-form-item-description"
          aria-expanded="false"
          aria-haspopup="dialog"
          aria-invalid="false"
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 w-[240px] pl-3 text-left font-normal"
          data-state="closed"
          id=":r7:-form-item"
          type="button"
        >
          September 25th, 2025
          <svg
            class="lucide lucide-calendar ml-auto h-4 w-4 opacity-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="18"
              rx="2"
              ry="2"
              width="18"
              x="3"
              y="4"
            />
            <line
              x1="16"
              x2="16"
              y1="2"
              y2="6"
            />
            <line
              x1="8"
              x2="8"
              y1="2"
              y2="6"
            />
            <line
              x1="3"
              x2="21"
              y1="10"
              y2="10"
            />
          </svg>
        </button>
      </div>
      <div
        class="space-y-2 max-w-[240px]"
      >
        <label
          class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 w-fit"
          for=":r9:-form-item"
        >
          Guests
        </label>
        <input
          aria-describedby=":r9:-form-item-description"
          aria-invalid="false"
          class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50"
          id=":r9:-form-item"
          name="quantityGuests"
          placeholder="number of guests"
          type="number"
          value="3"
        />
      </div>
      <div
        class="flex flex-col gap-4 mt-4"
      >
        <h4
          class="text-xl font-semibold tracking-tight"
        >
          Booking informations
        </h4>
        <div
          class="flex gap-6 justify-between"
        >
          <span
            class="underline"
          >
            $
            400
             x 
            5
             nights
          </span>
          <span>
            $
            2000
          </span>
        </div>
        <div
          class="flex gap-6 justify-between"
        >
          <span
            class="underline"
          >
            $
            50
             x 
            3
             guests
          </span>
          <span>
            $
            150
          </span>
        </div>
        <div
          class="shrink-0 bg-border h-[1px] w-full max-w-[200px] mx-auto"
          data-orientation="horizontal"
          role="none"
        />
        <div
          class="flex gap-6 justify-between"
        >
          <span
            class="text-lg font-semibold"
          >
            Total
          </span>
          <span
            class="text-lg font-bold"
          >
            $
            2150
          </span>
        </div>
      </div>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
        type="submit"
      >
        Save
      </button>
    </form>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 mt-4 w-full"
    >
      Cancel
    </button>
  </div>
</body>
`;

exports[`BookingForm GIVEN a BookingForm
      WHEN it's create
      THEN should NOT bring fields filled 1`] = `
<body>
  <div>
    <form
      class="flex flex-col gap-4"
    >
      <div
        class="space-y-2 flex flex-col"
      >
        <label
          class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 w-fit"
          for=":ra:-form-item"
        >
          Check-in
        </label>
        <button
          aria-controls="radix-:rb:"
          aria-describedby=":ra:-form-item-description"
          aria-expanded="false"
          aria-haspopup="dialog"
          aria-invalid="false"
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 w-[240px] pl-3 text-left font-normal text-muted-foreground"
          data-state="closed"
          id=":ra:-form-item"
          type="button"
        >
          <span>
            Pick a date to your check-in
          </span>
          <svg
            class="lucide lucide-calendar ml-auto h-4 w-4 opacity-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="18"
              rx="2"
              ry="2"
              width="18"
              x="3"
              y="4"
            />
            <line
              x1="16"
              x2="16"
              y1="2"
              y2="6"
            />
            <line
              x1="8"
              x2="8"
              y1="2"
              y2="6"
            />
            <line
              x1="3"
              x2="21"
              y1="10"
              y2="10"
            />
          </svg>
        </button>
      </div>
      <div
        class="space-y-2 flex flex-col"
      >
        <label
          class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 w-fit"
          for=":rc:-form-item"
        >
          Check-out
        </label>
        <button
          aria-controls="radix-:rd:"
          aria-describedby=":rc:-form-item-description"
          aria-expanded="false"
          aria-haspopup="dialog"
          aria-invalid="false"
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 w-[240px] pl-3 text-left font-normal text-muted-foreground"
          data-state="closed"
          disabled=""
          id=":rc:-form-item"
          type="button"
        >
          <span>
            Pick a date to your check-out
          </span>
          <svg
            class="lucide lucide-calendar ml-auto h-4 w-4 opacity-50"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="18"
              rx="2"
              ry="2"
              width="18"
              x="3"
              y="4"
            />
            <line
              x1="16"
              x2="16"
              y1="2"
              y2="6"
            />
            <line
              x1="8"
              x2="8"
              y1="2"
              y2="6"
            />
            <line
              x1="3"
              x2="21"
              y1="10"
              y2="10"
            />
          </svg>
        </button>
      </div>
      <div
        class="space-y-2 max-w-[240px]"
      >
        <label
          class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70 w-fit"
          for=":re:-form-item"
        >
          Guests
        </label>
        <input
          aria-describedby=":re:-form-item-description"
          aria-invalid="false"
          class="flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-sm shadow-sm transition-colors file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50"
          id=":re:-form-item"
          name="quantityGuests"
          placeholder="number of guests"
          type="number"
          value=""
        />
      </div>
      <div
        class="flex flex-col gap-4 mt-4"
      >
        <h4
          class="text-xl font-semibold tracking-tight"
        >
          Booking informations
        </h4>
        <div
          class="flex gap-6 justify-between"
        >
          <span
            class="underline"
          >
            $
            400
             x 
            0
             nights
          </span>
          <span>
            $
            0
          </span>
        </div>
        <div
          class="flex gap-6 justify-between"
        >
          <span
            class="underline"
          >
            $
            50
             x 
            0
             guests
          </span>
          <span>
            $
            0
          </span>
        </div>
        <div
          class="shrink-0 bg-border h-[1px] w-full max-w-[200px] mx-auto"
          data-orientation="horizontal"
          role="none"
        />
        <div
          class="flex gap-6 justify-between"
        >
          <span
            class="text-lg font-semibold"
          >
            Total
          </span>
          <span
            class="text-lg font-bold"
          >
            $
            0
          </span>
        </div>
      </div>
      <button
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
        type="submit"
      >
        Book
      </button>
    </form>
    <button
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2 mt-4 w-full"
    >
      Cancel
    </button>
  </div>
</body>
`;
